### DISCUSSION TITLE
Are there any plans for storing messages encrypted at rest?

### DISCUSSION THREAD

**Discussion by smorks** [stalwartlabs/jmap-server#42] [Ideas]
From what I understand, all metadata is stored in RocksDB, but the messages themselves are stored unencrypted on the filesystem.

It would be nice to have the option to setup some sort of encryption key, maybe based on the user's password, so the message can be encrypted when written to disk. I'm sure there would be a lot of work involved with this, such as having to re-encrypt all the messages when a password is changed, and some performance overhead with encrypting/decrypting on the fly when reading/writing messages, and many more things that i'm sure i haven't thought about.

for reference, dovecot has the [mail-crypt-plugin](https://doc.dovecot.org/configuration_manual/mail_crypt_plugin/), not sure if it's design would be helpful at all?

**Comment by dumblob** [comment]
Do you mean this: https://github.com/facebook/rocksdb/commit/51778612c9cf4cb842eed10f270ec0ed29ff22f1 ?

**Comment by mdecimus** [comment]
Implementing encryption for raw messages should not be that complex, the only issue is being able to search for messages as a full copy of the body and subject needs to be stored in the database. That being said, encryption at rest is something that is planned after the JMAP server is refactored.

**Reply by smorks** [reply]
great! thanks for the response. love the work you've done on the jmap/imap/smtp servers.
